masterclass: Initial PR for balenaCLI advanced masterclass (#1)
balena-io/balena-cli-advanced-masterclass: README.md

balenaCLI Advanced Masterclass
==============================

# Prerequisite Classes

This masterclass builds upon knowledge that has been taught in previous classes.

> To gain the most from this masterclass, we reommend that you first undertake

Reviewer: ```suggestion
To gain the most from this masterclass, we recommend that you first complete
```

Author: Done

introducing you to additional features that can be used to gain finer control
over the provisioning, deployment and management of devices.

# Hardware and Software Requirements

The hardware and software requirements for the masterclass. This should be a
list of these items, along with hyperlinks to relevant product pages or
installation guides.

* A balena supported device, such as a [balenaFin 1.1](https://store.balena.io/collections/developer-kit/products/balenafin-v1-1-0-developer-kit),
	[Raspberry Pi 3](https://www.raspberrypi.org/products/raspberry-pi-3-model-b/)
	or [Intel NUC](https://www.intel.co.uk/content/www/uk/en/products/boards-kits/nuc.html). If you don't have a device, you can emulate an Intel NUC by
	installing VirtualBox and following [this guide](https://www.balena.io/blog/no-hardware-use-virtualbox/)
* A suitable text editor for developing code on your development platform (eg.
    [Visual Code](https://code.visualstudio.com/))
* A suitable shell environment for command execution (such as `bash`)
* A [balenaCloud](https://www.balena.io/) account
* A familiarity with [Dockerfiles](https://docs.docker.com/engine/reference/builder/)

> * A [NodeJS](https://nodejs.org/en/) installation, including [NPM](https://www.npmjs.com/get-npm). The use of [`nvm`](https://github.com/nvm-sh/nvm)

Reviewer 2: I feel like we should rather recommend the installers and avoid the node + Npm install path for the CLI.

Author: Changed to fall inline with what I've written in the other masterclass.

environment and the `stagingCliApp` application:
```
$ BALENARC_DATA_DIRECTORY=. balena join 192.168.1.171 --application stagingCliApp
? Check for updates every X minutes 10
[Success] Device successfully joined balena-staging.com!
```
We can now check the devices on the staging environment to ensure it's joined
successfully:
```
$ BALENARC_DATA_DIRECTORY=. balena apps
BALENARC_DATA_DIRECTORY=. balena devices
ID     UUID    DEVICE NAME      DEVICE TYPE  APPLICATION NAME  STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
876542 3456789 purple-snowflake fincm3       stagingCliApp     Idle   true      9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-staging.com/devices/3456789345678934567893456789/summary
```
As can be seen, it's been given a new name and UUID.

If we hadn't specified the application to join, we would have seen an
interactive list of all the applications on the staging environment:
```
$ BALENARC_DATA_DIRECTORY=. balena join 192.168.1.171
? Select application
❯ heds/artik530
  heds/orangepi
  heds/stagingcliapp
  heds/switchapp
  heds/vpn-changed
```
Note that the applications listed are not all of the `fincm3` device type, but

> share the common `armv7hf` architecture. Be very careful to select an

Reviewer 2: I think there was a newer update that would allow moving into armv6 fleets too. I think thodoris add the mix capability fleets to the CLI.

Author: Clarified this.

where a device may require tracking, as a specific device UUID can then be
associated with a specific customer order, for example. This ensures that
a customer then receives a device that already has information available
for it, for support, application updates, etc.

The following exercises will show you how to both preload and preregister
devices.

## 5.5 Preloading a Device Image

Note that `balena preload` actually uses a Docker container to carry out the
actual preloading. This is to allow the system to be portable and run under
Linux, macOS and Windows. As such, you'll need to ensure you've installed
Docker on your development machine before going any further.

We'll take the current `cliApp` application and preload it the
`balena-fin-image.img` image we earlier configured and provisioned our device
with. The `balena preload` command has a large number of switch options for
catering with different situations, including the ability to use a particular
release. However, for this exercise, we'll simply use the latest version of
the code we previously pushed to `cliApp`. In a terminal, execute the following:

> >>>>> NOTE THAT THIS DOESN'T WORK ON MACOS, AS WE SEE AN ERROR WHEN ATTEMPTING

Reviewer 2: Is this error because docker for MacOS doesn't support AUFS?

Author: I need to look into this in a bit more detail. If we're assuming AUFS, then definitely, but I don't think I've ever managed to get it working under macOS.

Author: Yeah, I found out why. When Docker did a major update change, it wiped previously shared directories. I've had to re-add my `/Work` directory. Sigh.

balenaCLI Advanced Masterclass
==============================

# Prerequisite Classes

This masterclass builds upon knowledge that has been taught in previous classes.
To gain the most from this masterclass, we reommend that you first undertake
the following masterclasses:

* [balenaCLI Masterclass](https://github.com/balena-io-projects/balena-cli-masterclass)

**Masterclass Type:** Core
**Maximum Expected Time To Complete:** 120 minutes

# Introduction

> The balena Command Line Interface (balenaCLI) utility consists of a number

Reviewer: ```suggestion
The balena Command Line Interface (balena-cli) utility consists of a number
```

Reviewer: This change is just for consistency with how we name/call the product (https://github.com/balena-io/balena-cli)

Author: Hmm, I'd rather change this to 'balena CLI' as how we refer to it in the docs, but not the package name. I know the package name is very similar, but 'balena CLI' refers to the CLI method of operating with balena, rather than the package name (and if we ever changed the package name, it's still valid).

Reviewer: As long as it's consistent, that's fine with me!

of commands that allow a user to develop, deploy and manage balena applications
and devices.

> The previous balenaCLI masterclass covered some of the most common techniques

Reviewer: ibid.

> and topics related to balenaCLI. This masterclass aims to build on top of that,

Reviewer: ibid.

Author: Shortened.

introducing you to additional features that can be used to gain finer control
over the provisioning, deployment and management of devices.

# Hardware and Software Requirements

> The hardware and software requirements for the masterclass. This should be a

Reviewer: Is this leftover template?

Author: Very good spot. Yes!

12346 appTwo            iot2000      0              0
```

# 1.2 Configuration File

Should you just wish to use balenaCLI without specifying the environment to use
in an environment variable, you can use a configuration file instead.

By default, balenaCLI looks for a configuration file in the user's home
directory. We can demonstrate configuring balenaCLI for an alternative
environment by creating a new file as `~/.balenarc.yml` and then filling
it with the following information:
```
balenaUrl: 'balena-staging.com'
```
Now try listing the apps from the staging environment again:
```
$ balena apps
ID    APP NAME          DEVICE TYPE  ONLINE DEVICES DEVICE COUNT
12345 appOne            intel-nuc    0              0
12346 appTwo            iot2000      0              0
```
Two things have happened here. The first is that balenaCLI has found and used
the new balena environment URL from the `~/.balenarc.yml` configuration file.
The second is that it has used the token previously retrieved from logging in
to the staging environment using the environment variable.

Token files are saved separately and can be found in the user's home directory

> here: `~/.balena/token`.

Reviewer: Is this true in non *nix environments? I've no idea how Windows file hierarchies work.

Author: For WSL, yes. For normal shell, no idea. I've clarified this.

environments (for example when testing new features available on staging but
not production).

To enable this, balenaCLI includes the ability to use a configuration file in
the current working directory (CWD) to determine which environment to use,
as well as an environment variable to determine where to store the token
and data.

However, when doing so, it merges together *all* configurations that it finds,
with any files in the user's home directory taking precedence. To demonstrate
using different configuration files for different environment, first remove
the configuration files from your home directory so these do not override the
configurations we will create:
```
$ rm ~/.balenarc.yml
$ rm -rf ~/.balena
```

Now, in your home directory, create two directories called `balenaProduction`
and `balenaStaging`, and then fill in separate configuration files in each
directory. The following commands will do this for you:
```
$ mkdir -p ~/balenaProduction
$ mkdir -p ~/balenaStaging
$ echo "balenaUrl: 'balena-staging.com'" > ~/balenaStaging/balenarc.yml
$ echo "balenaUrl: 'balena-cloud.com'" > ~/balenaProduction/balenarc.yml
```

> Note that we do *not* prefix the `balenarc.yml` file with a `.` to make it

Reviewer: ```suggestion
Note that we do *not* prefix the `balenarc.yml` file with a `.` to avoid hiding it
```

 _            _
| |__   __ _ | |  ____  _ __    __ _
| '_ \ / _` || | / __ \| '_ \  / _` |
| |_) | (_) || ||  ___/| | | || (_) |
|_.__/ \__,_||_| \____/|_| |_| \__,_|

Logging in to balena-staging.com
? How would you like to login? (Use arrow keys)
❯ Web authorization (recommended)
  Credentials
  Authentication token
  I don't have a balena account!
```
This time, a `token` file was created in the `~/balenaStaging` directory.

You can now switch between environments by changing directory to the one
with the relevant `balenarc.yml` and `token` file, which will allow you
to use either environment without any further authentication:
```
$ cd ~/balenaStaging
$ BALENARC_DATA_DIRECTORY=. balena apps
ID    APP NAME          DEVICE TYPE  ONLINE DEVICES DEVICE COUNT
12345 appOne            intel-nuc    0              0
12346 appTwo            iot2000      0              0
$ cd ~/balenaProduction
$ BALENARC_DATA_DIRECTORY=. balena apps
ID      APP NAME         DEVICE TYPE      ONLINE DEVICES DEVICE COUNT
54321   artik530         artik533s        0              0
54322   bob              intel-nuc        0              0
54323   bbbtest          beaglebone-black 0              0
```

> We'll be using the two separate environments in the next set of exercises

Reviewer: @wrboyce has some nice bash aliases that help managing multiple envs with the CLI, perhaps that's worth including?

Reviewer 3: Not sure they're good for public consumption really ;)

Author: Yeah, I'd rather stick to actual commands here rather than 'bash magic'. If users want to do scripting that's fine, but we really shouldn't be adding scripts here.

Reviewer 3: Would be a worthwhile hack friday to add multiple environment support to `balena-cli` mind. I envisage something similar to how aws cli tooling works.

Author: @wrboyce Dude, that would be utterly awesome

test environment using openBalena and now want to move from your test
environment to a the production environment of balenaCloud).

The following exercises will show you how to carry this out.

## 2.1 Moving Devices between Applications

Moving a device to another application on the same environment is extremely
easy. To demonstrate this, first create a new application for the balenaFin:
```
$ BALENARC_DATA_DIRECTORY=. balena app create altApp --type fincm3
Application created: altApp (fincm3, id 987654)
```
We should already have a device connecting to our previous 'cliApp' (from the
previous
[balena CLI Masterclass](https://github.com/balena-io-projects/balena-cli-masterclass)):
```
$ BALENARC_DATA_DIRECTORY=. balena devices
ID      UUID    DEVICE NAME          DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE
1234556 7654321 weathered-wildflower fincm3          cliApp           Idle   false     9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-cloud.io/devices/1234567890abcdef/summary
```
To interactively determine which application to move a device to, simply use
its UUID with the `balena device move` command:
```

> $ BALENARC_DATA_DIRECTORY=. balena device move 7654321

Reviewer: misread this as the device ID, opened https://github.com/balena-io/balena-cli/issues/1457 because that behavior was confusing for me.

Author: Yeah, UUID is pretty much used for every command... except where it's not. Issue's a good call.

name:
```
$ BALENARC_DATA_DIRECTORY=. balena device move 94095f8 --application altApp
94095f8 was moved to altApp
```
A call of `balena device` specifying the UUID of the moved device will now
show it is owned by the specified application:
```
$ BALENARC_DATA_DIRECTORY=. balena device 7654321
== WEATHERED WILDFLOWER
ID:                 1693707
DEVICE TYPE:        fincm3
STATUS:             idle
IS ONLINE:          true
IP ADDRESS:         192.168.1.171 192.168.1.169
APPLICATION NAME:   altApp
UUID:               76543217654321765432176543217654
COMMIT:             7efbc95825641b6482742a54c8e74010
SUPERVISOR VERSION: 9.15.7
IS WEB ACCESSIBLE:  false
OS VERSION:         balenaOS 2.38.0+rev1
DASHBOARD URL:      https://dashboard.balena-cloud.com/devices/76543217654321765432176543217654/summary
```
The Supervisor on the device will remove any previously running services, as
well as their images and any volumes associated with them and download the
images associated with the new application before starting them.

## 2.2 Moving Devices between Environments

Moving a device between balena environments is slightly more involved,

> and alters depending on whether you're using a device running a development

Reviewer: ```suggestion
and differs depending on whether you're using a device running a development
```

or production image.

### 2.2.1 Development Devices

For a device running a development image, you can use `balena leave` and
`balena join` to carry this out.

First, provision a device using a development image. You can do this using
the balenaCloud dashboard and downloading a 'Development' edition image from
the `cliApp` application page. Provision your balenaFin with this image

> using either balenaEtcher or balenaCLI.

Reviewer: Real question: how can you provision a fin with the CLI? I've never heard of this, and am super curious. Maybe i missed this in the `beginners` masterclass?

Reviewer: ah, ha, farther down (perhaps worth linking to internally)?

Author: Yes, it's further down, I'll link.

Once the device is provisioned and has connected to the balena network,
discover its hostname or IP address by using `balena devices`:
```
$ BALENARC_DATA_DIRECTORY=. balena devices
ID      UUID    DEVICE NAME      DEVICE TYPE     APPLICATION NAME STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
1234567 7654321 little-paper     fincm3          cliApp           Idle   true      9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-cloud.com/devices/76543217654321765432176543217654/summary

$ BALENARC_DATA_DIRECTORY=. balena device 7654321
== LITTLE PAPER
ID:                 1234567
DEVICE TYPE:        fincm3
STATUS:             idle
IS ONLINE:          true
IP ADDRESS:         192.168.1.171
APPLICATION NAME:   cliApp
UUID:               76543217654321765432176543217654
COMMIT:             7efbc95825641b6482742a54c8e74010
SUPERVISOR VERSION: 9.15.7
IS WEB ACCESSIBLE:  false
OS VERSION:         balenaOS 2.38.0+rev1
DASHBOARD URL:      https://dashboard.balena-cloud.com/devices/76543217654321765432176543217654/summary
```
As we now have the local IP address for it, we can use this to call the command
to leave the balenaCloud environment:
```

> $ BALENARC_DATA_DIRECTORY=. balena leave 192.168.1.171

Reviewer: Not important, but I tried this with a NUC VM running via the VBox NAT networking stack and the command just hung.

Author: Interesting, tried this on both a Pi4 and Fin and it worked. What did the Supervisor on the VM say?

BALENARC_DATA_DIRECTORY=. balena devices
ID     UUID    DEVICE NAME      DEVICE TYPE  APPLICATION NAME  STATUS IS ONLINE SUPERVISOR VERSION OS VERSION           DASHBOARD URL
876542 3456789 purple-snowflake fincm3       stagingCliApp     Idle   true      9.15.7             balenaOS 2.38.0+rev1 https://dashboard.balena-staging.com/devices/3456789345678934567893456789/summary
```
As can be seen, it's been given a new name and UUID.

If we hadn't specified the application to join, we would have seen an
interactive list of all the applications on the staging environment:
```
$ BALENARC_DATA_DIRECTORY=. balena join 192.168.1.171
? Select application
❯ heds/artik530
  heds/orangepi
  heds/stagingcliapp
  heds/switchapp
  heds/vpn-changed
```
Note that the applications listed are not all of the `fincm3` device type, but
share a common architecture, so this could be `aarch64`, `armv7hf` or `armv6`
for the Fin. You should be very careful when selecting an application of a
different device type, as you may find issues when the device attempts to run
the application code for a device without the same peripherals or system-on-chip
layout.

### 2.2.2 Production Devices

> >>> THIS WILL NOT WORK FOR A PRODUCTION DEVICE, BECAUSE AS SOON AS YOU USE

Reviewer: maybe worth linking to any relevant issue(s) in all bug sections?

Author: Yup

Author: Done

[Logs]    [9/17/2019, 11:06:39 AM] Installed service 'main sha256:4e11240a66e7e86f38f0ccf9a806a079c4181d446bc026d6b62b83981781737b'
[Logs]    [9/17/2019, 11:06:39 AM] Starting service 'main sha256:4e11240a66e7e86f38f0ccf9a806a079c4181d446bc026d6b62b83981781737b'
[Logs]    [9/17/2019, 11:06:40 AM] Started service 'main sha256:4e11240a66e7e86f38f0ccf9a806a079c4181d446bc026d6b62b83981781737b'
[Logs]    [9/17/2019, 11:06:42 AM] [main]
[Logs]    [9/17/2019, 11:06:42 AM] [main] > balena-cli-advanced-masterclass@1.0.0 start /usr/src/app
[Logs]    [9/17/2019, 11:06:42 AM] [main] > node src/helloworld.js; sleep infinity
[Logs]    [9/17/2019, 11:06:42 AM] [main]
[Logs]    [9/17/2019, 11:06:42 AM] [main] ---> LOG_DEBUG not set, no debug mode
[Logs]    [9/17/2019, 11:06:42 AM] [main] Hello world!
```
We're back to no debug logs again!

The balenaCLI also allows you to rename environment variables, as well as set
and remove them for specific devices instead of an entire application fleet.
For device-based variables, most commands take an extra switch, `--device`
which allows you to specify the UUID of the device you wish to make the
variable change for.

# 5. Preloading and Preregistering

> Whilst for many cases provisoining a device, moving it to an installation

Reviewer: ```suggestion
Whilst for many cases provisioning a device, moving it to an installation
```
